tests/test_engine_db.py

```diff
@@ -0,0 +1,193 @@
+        (ObjectStatus.INITIAL, "New"),
+        (ObjectStatus.COMPLETED, "Done"),
+        (ObjectStatus.HALTED, "Need action"),
+        (ObjectStatus.RUNNING, "In process"),
+        (ObjectStatus.WAITING, "Waiting"),
+        (ObjectStatus.ERROR, "Error"),
+    ))
+    def test_object_status_name_returns_correct_name(self, status, name):
+        assert ObjectStatus.name(status) == name
+
+
+class TestWorkflowEngineDb(object):
+
+    def setup_method(self, method):
+        self.dummy_db_obj = mock.Mock(spec=DummyDbObj())
+        self.wfe = DbWorkflowEngine(self.dummy_db_obj)
+        self.data = ['one', 'two', 'three', 'four', 'five']
+        self.tokens = [mock.Mock(spec=FakeToken(x)) for x in self.data]
+
+    def teardown_method(self, method):
+        pass
+
+    @mock.patch.object(TransitionActions, 'HaltProcessing')
+    def test_halt_processing_calls_parent(self, mock_HaltProcessing):
+        self.wfe.callbacks.add_many([
+            m('mouse'),
+            lambda obj, eng: eng.halt()
+        ])
+        self.wfe.process(self.tokens)
+
+        assert mock_HaltProcessing.call_count == len(self.data)
+        for args_list in mock_HaltProcessing.call_args_list:
+            args_list = args_list[0]
+            assert isinstance(args_list[0], FakeToken)
+            assert isinstance(args_list[1], DbWorkflowEngine)
+            assert isinstance(args_list[2], Iterable)
+            assert isinstance(args_list[3], HaltProcessing)
+
+    def test_halt_processing_saves_eng_and_obj(self):
+        self.wfe.callbacks.add_many([
+            lambda obj, eng: eng.halt('please wait')
+        ])
+        with pytest.raises(HaltProcessing):
+            self.wfe.process(self.tokens)
+
+        token = self.tokens[0]
+
+        assert token.save.call_count == 2
+        assert token.save.call_args_list[0][1]['version'] == ObjectStatus.RUNNING
+        assert token.save.call_args_list[1][1]['version'] == ObjectStatus.WAITING  # (Halt is WAITING if not action is set)
+
+    def test_halt_processing_saves_correct_statuses(self):
+        self.wfe.callbacks.add_many([
+            lambda obj, eng: eng.halt('please wait')
+        ])
+        with pytest.raises(HaltProcessing):
+            self.wfe.process(self.tokens)
+
+        # Token saved
+        token = self.tokens[0]
+        assert token.save.call_count == 2
+        assert token.save.call_args_list[0][1]['version'] == ObjectStatus.RUNNING
+        assert token.save.call_args_list[1][1]['version'] == ObjectStatus.WAITING  # (Halt is WAITING if not action is set)
+
+        # Engine saved
+        assert self.dummy_db_obj.save.call_count == 3
+        assert self.dummy_db_obj.save.call_args_list[0][0] == (WorkflowStatus.NEW, )
+        assert self.dummy_db_obj.save.call_args_list[1][0] == (WorkflowStatus.RUNNING, )
+        assert self.dummy_db_obj.save.call_args_list[2][0] == (WorkflowStatus.HALTED, )
+
+    # TODO: grep for '_counter_' in engine_db and test it
```

Comment:
Regarding these counters,

* I doubt they should be part of the interface. They are a burden to increment all the time and I don't think they're useful enough.
* In the model of invenio I just shadowed them completely with python properties because they can be inferred: https://github.com/dset0x/invenio/commit/25657974bf35e6dc88ce4898d2caa3780c8b830b